metadata-service/auth-impl/src/main/java/com/datahub/authentication/user/NativeUserService.java

```diff
@@ -0,0 +1,292 @@
+package com.datahub.authentication.user;
+
+import com.datahub.authentication.Authentication;
+import com.linkedin.common.AuditStamp;
+import com.linkedin.common.urn.CorpuserUrn;
+import com.linkedin.common.urn.Urn;
+import com.linkedin.entity.client.EntityClient;
+import com.linkedin.events.metadata.ChangeType;
+import com.linkedin.identity.CorpUserCredentials;
+import com.linkedin.identity.CorpUserInfo;
+import com.linkedin.identity.CorpUserStatus;
+import com.linkedin.identity.InviteToken;
+import com.linkedin.metadata.entity.EntityService;
+import com.linkedin.metadata.secret.SecretService;
+import com.linkedin.metadata.utils.GenericRecordUtils;
+import com.linkedin.mxe.MetadataChangeProposal;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.security.MessageDigest;
+import java.security.SecureRandom;
+import java.time.Instant;
+import java.util.Base64;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+import lombok.extern.slf4j.Slf4j;
+
+import static com.linkedin.metadata.Constants.*;
+
+
+/**
+ * Service responsible for creating, updating and authenticating native DataHub users.
+ */
+@Slf4j
+public class NativeUserService {
+  private static final int LOWERCASE_ASCII_START = 97;
+  private static final int LOWERCASE_ASCII_END = 122;
+  private static final int INVITE_TOKEN_LENGTH = 32;
+  private static final int SALT_TOKEN_LENGTH = 16;
+  private static final int PASSWORD_RESET_TOKEN_LENGTH = 32;
+  private static final String HASHING_ALGORITHM = "SHA-256";
+  private static final long ONE_DAY_MILLIS = TimeUnit.DAYS.toMillis(1);
+
+  private final EntityService _entityService;
+  private final EntityClient _entityClient;
+  private final SecretService _secretService;
+  private final SecureRandom _secureRandom;
+  private final MessageDigest _messageDigest;
+
+  public NativeUserService(@Nonnull EntityService entityService, @Nonnull EntityClient entityClient, @Nonnull SecretService secretService)
+      throws Exception {
+    Objects.requireNonNull(entityService, "EntityService must not be null!");
```

Comment:
Generally, you should place the exact name of the arg in the message..
For example:

entityService must not be null
entityClient must not be null
secretService must not be null

Reply:
Because these are lower-level checks - not intended to convey higher level meaning!

Reply:
Makes sense!